firedrake/preconditioners/hypre_ads.py

```diff
@@ -0,0 +1,60 @@
+from firedrake.preconditioners.base import PCBase
+from firedrake.petsc import PETSc
+from firedrake.functionspace import FunctionSpace, VectorFunctionSpace
+from firedrake.ufl_expr import TestFunction
+from firedrake.interpolation import Interpolator
+from firedrake.dmhooks import get_function_space
+from ufl import grad, curl
+
+__all__ = ("HypreADS",)
+
+
+class HypreADS(PCBase):
+    def initialize(self, obj):
+        A, P = obj.getOperators()
+        prefix = obj.getOptionsPrefix()
+        V = get_function_space(obj.getDM())
+        mesh = V.mesh()
+
+        family = str(V.ufl_element().family())
+        degree = V.ufl_element().degree()
+        if family != 'Raviart-Thomas' or degree != 1:
+            raise ValueError("Hypre ADS requires lowest order RT elements! (not %s of degree %d)" % (family, degree))
+
+        P1 = FunctionSpace(mesh, "Lagrange", 1)
+        NC1 = FunctionSpace(mesh, "N1curl", 1)
+        # DiscreteGradient
+        G = Interpolator(grad(TestFunction(P1)), NC1).callable().handle
+        # DiscreteCurl
+        C = Interpolator(curl(TestFunction(NC1)), V).callable().handle
+
+        pc = PETSc.PC().create(comm=obj.comm)
+        pc.incrementTabLevel(1, parent=obj)
+        pc.setOptionsPrefix(prefix + "hypre_ads_")
+        pc.setOperators(A, P)
+
+        pc.setType('hypre')
+        pc.setHYPREType('ads')
+        pc.setHYPREDiscreteGradient(G)
+        pc.setHYPREDiscreteCurl(C)
+        V = VectorFunctionSpace(mesh, "Lagrange", 1)
+        linear_coordinates = Interpolator(SpatialCoordinate(mesh), V).dat.data_ro.copy()
```

Comment:
Ah sorry, we have to import `SpatialCoordinate` from `ufl` at the top of the file for this to work.

Reply:
OK, I will resolve it right now.

Reply:
I have got lost. Could you help me?

Reply:
The new changes look to be correct. And the only failing test I now see is unrelated to your changes (and is a problem we are trying solve separately).

Reply:
Many thanks.